jib-core/src/main/java/com/google/cloud/tools/jib/ncache/DefaultCacheStorageReader.java

```diff
@@ -58,4 +60,46 @@
       return layerDigests;
     }
   }
+
+  /**
+   * Retrieves the {@link CacheEntry} for the layer with digest {@code layerDigest}.
+   *
+   * @param layerDigest the layer digest
+   * @return the {@link CacheEntry} referenced by the layer digest
+   * @throws CacheCorruptedException if the cache was found to be corrupted
+   * @throws IOException if an I/O exception occurs
+   */
+  Optional<CacheEntry> retrieve(DescriptorDigest layerDigest)
+      throws IOException, CacheCorruptedException {
+    Path layerDirectory = defaultCacheStorageFiles.getLayerDirectory(layerDigest);
+    if (!Files.exists(layerDirectory)) {
+      return Optional.empty();
+    }
+
+    DefaultCacheEntry.Builder cacheEntryBuilder =
+        DefaultCacheEntry.builder().setLayerDigest(layerDigest);
+
+    try (Stream<Path> filesInLayerDirectory = Files.list(layerDirectory)) {
+      for (Path fileInLayerDirectory : filesInLayerDirectory.collect(Collectors.toList())) {
+        if (DefaultCacheStorageFiles.isLayerFile(fileInLayerDirectory)) {
+          if (cacheEntryBuilder.getLayerBlob().isPresent()) {
```

Comment:
I'm a little confused, so bear with me: `CacheEntry` seem kind of like a union type, so this code feels weird. Should it either build `metadataCacheEntry` or `layerCacheEntry`? Can a cache entry contain both `metadata` and `layer`? If not, can we have multiple metadata files? can line 98 be hit more than once?

Reply:
Yea, `CacheEntry` is a union type of either just `layer` data or both `layer` and `metadata` data (the former being base image layers and the latter being application layers). Line 98 won't be hit more than once because the metadata file is named by exact name, though this code should probably not make that assumption. I'll change it to not make that assumption.

Reply:
Are they functionally different in any other part of the code? Should they be handled as different entry types?

Reply:
When being read, they are not functionally different. The only use for the metadata currently is to check the last modified time when trying to see if a layer needs rebuilding (select the layer by the selector, then check if the files have any modified times greater than that stored in the metadata).